File: web/src/core/detection/adaptiveThreshold.ts
/**
 * Robust, adaptive amplitude threshold — port of detection/adaptive_threshold.py.
 * Median/MAD rather than mean/std: robust to the very outliers we're
 * thresholding against.
 */

export const MAD_TO_SIGMA = 1.4826;

/** Matches numpy.median exactly: average of the two middle elements for an
 * even-length array (numpy does NOT use nearest-rank / "lower" median). */
export function median(values: number[]): number {
  if (values.length === 0) return 0;
  const sorted = [...values].sort((a, b) => a - b);
  const mid = sorted.length >> 1;
  return sorted.length % 2 === 1 ? sorted[mid] : (sorted[mid - 1] + sorted[mid]) / 2;
}

export function medianAbsoluteDeviation(values: number[], med: number): number {
  const deviations = values.map((v) => Math.abs(v - med));
  return median(deviations);
}

/** Fixed-capacity FIFO buffer, mirroring Python's collections.deque(maxlen=...). */
export class RingBuffer {
  private buf: number[] = [];
  private readonly maxLen: number;

  constructor(maxLen: number) {
    this.maxLen = maxLen;
  }

  push(value: number): void {
    this.buf.push(value);
    if (this.buf.length > this.maxLen) this.buf.shift();
  }

  get values(): number[] {
    return this.buf;
  }

  get length(): number {
    return this.buf.length;
  }
}

export class AdaptiveThreshold {
  private readonly buffer: RingBuffer;
  private readonly madMultiplier: number;
  private readonly updateIntervalSamples: number;
  private readonly minThreshold: number;
  private samplesSinceUpdate = 0;
  private medianValue = 0;
  private madValue = 0;
  private thresholdValue: number;

  constructor(fsHz: number, windowS = 10.0, madMultiplier = 4.0, updateIntervalSamples = 32, minThreshold = 1e-6) {
    this.madMultiplier = madMultiplier;
    this.updateIntervalSamples = updateIntervalSamples;
    this.minThreshold = minThreshold;
    const maxLen = Math.max(Math.floor(windowS * fsHz), 8);
    this.buffer = new RingBuffer(maxLen);
    this.thresholdValue = minThreshold;
  }

  /** Feed one (rectified/abs) filtered sample; returns the current threshold. */
  update(absSampleValue: number, inCandidate = false): number {
    if (!inCandidate) {
      this.buffer.push(absSampleValue);
      this.samplesSinceUpdate += 1;
    }

    if (this.samplesSinceUpdate >= this.updateIntervalSamples && this.buffer.length >= 8) {
      const values = this.buffer.values;
      this.medianValue = median(values);
      this.madValue = medianAbsoluteDeviation(values, this.medianValue);
      this.thresholdValue = Math.max(
        this.medianValue + this.madMultiplier * MAD_TO_SIGMA * this.madValue,
        this.minThreshold,
      );
      this.samplesSinceUpdate = 0;
    }

    return this.thresholdValue;
  }

  get threshold(): number {
    return this.thresholdValue;
  }

  get noiseFloorMedian(): number {
    return this.medianValue;
  }

  get noiseFloorMad(): number {
    return this.madValue;
  }

  /** Bulk-seed the buffer (used right after calibration) so the threshold
   * is meaningful immediately instead of ramping up from zero. */
  seed(samples: Float64Array | number[]): void {
    for (const s of samples) this.buffer.push(Math.abs(s));
    this.samplesSinceUpdate = this.updateIntervalSamples;
    this.update(0, true);
  }
}

File: src/bcihand/detection/adaptive_threshold.py
"""Robust, adaptive amplitude threshold for candidate detection.

Uses median / MAD (median absolute deviation) rather than mean/std, per the
project brief: robust statistics resist being dragged around by the very
outliers (spikes, artifacts) we're trying to threshold against.

NOTE for embedded port: recomputing an exact median/MAD over a multi-second
rolling buffer is O(n log n) and too expensive for a tight MCU loop. On the
ESP32 this should become a cheaper running-percentile approximation (e.g. a
stochastic/streaming quantile estimator) — tracked in docs/SIGNAL_PIPELINE.md.
On the PC this exact form is fine and gives us a correctness baseline to
approximate against later.
"""
from __future__ import annotations

from collections import deque

import numpy as np

MAD_TO_SIGMA = 1.4826  # consistency constant for a normal distribution


class AdaptiveThreshold:
    def __init__(
        self,
        fs_hz: float,
        window_s: float = 10.0,
        mad_multiplier: float = 4.0,
        update_interval_samples: int = 32,
        min_threshold: float = 1e-6,
    ):
        self.fs_hz = fs_hz
        self.maxlen = max(int(window_s * fs_hz), 8)
        self.mad_multiplier = mad_multiplier
        self.update_interval_samples = update_interval_samples
        self.min_threshold = min_threshold

        self._buffer: deque[float] = deque(maxlen=self.maxlen)
        self._samples_since_update = 0
        self._median = 0.0
        self._mad = 0.0
        self._threshold = min_threshold

    def update(self, abs_sample_value: float, in_candidate: bool = False) -> float:
        """Feed one (rectified/abs) filtered sample; returns the current threshold.

        Samples flagged `in_candidate=True` (i.e. part of an active blink
        candidate) are excluded from the noise-floor estimate so genuine
        blinks don't inflate their own detection threshold.
        """
        if not in_candidate:
            self._buffer.append(abs_sample_value)
            self._samples_since_update += 1

        if self._samples_since_update >= self.update_interval_samples and len(self._buffer) >= 8:
            arr = np.fromiter(self._buffer, dtype=np.float64, count=len(self._buffer))
            self._median = float(np.median(arr))
            self._mad = float(np.median(np.abs(arr - self._median)))
            self._threshold = max(
                self._median + self.mad_multiplier * MAD_TO_SIGMA * self._mad,
                self.min_threshold,
            )
            self._samples_since_update = 0

        return self._threshold

    @property
    def threshold(self) -> float:
        return self._threshold

    @property
    def noise_floor_median(self) -> float:
        return self._median

    @property
    def noise_floor_mad(self) -> float:
        return self._mad

    def seed(self, samples: np.ndarray) -> None:
        """Bulk-seed the buffer (used right after calibration) so the threshold
        is meaningful immediately instead of ramping up from zero."""
        for s in samples:
            self._buffer.append(float(abs(s)))
        self._samples_since_update = self.update_interval_samples  # force recompute
        self.update(0.0, in_candidate=True)  # trigger recompute without adding a sample
